gms/gossiper.cc

```diff
@@ -1020,21 +1020,22 @@ std::set<inet_address> gossiper::get_live_members() const {
 
 std::set<inet_address> gossiper::get_live_token_owners() const {
     std::set<inet_address> token_owners;
-    for (auto& member : get_live_members()) {
-        auto es = get_endpoint_state_for_endpoint_ptr(member);
-        if (es && !is_dead_state(*es) && get_token_metadata_ptr()->is_member(member)) {
-            token_owners.insert(member);
+    auto normal_token_owners = get_token_metadata_ptr()->get_all_endpoints();
+    for (auto& node: normal_token_owners) {
+        if (is_alive(node)) {
+            token_owners.insert(node);
         }
     }
     return token_owners;
+
 }
 
 std::set<inet_address> gossiper::get_unreachable_token_owners() const {
     std::set<inet_address> token_owners;
-    for (auto&& x : _unreachable_endpoints) {
-        auto& endpoint = x.first;
-        if (get_token_metadata_ptr()->is_member(endpoint)) {
-            token_owners.insert(endpoint);
+    auto normal_token_owners = get_token_metadata_ptr()->get_all_endpoints();
+    for (auto& node: normal_token_owners) {
+        if (!is_alive(node)) {
+            token_owners.insert(node);
```

Comment:
nit: It's worth generalizing this into smth like

```c++
template <typename Fn>
std::set<inet_address> token_metadata::get_endpoints_filtered(Fn&& filter) const;

std::set<inet_address> gossiper::get_unreachable_token_owners() const {
    return get_token_metadata_ptr()->get_endpoints_filtered([this] (auto ep) { return !is_alive(ep); });
}
```

Reply:
I'd keep the change minimal. It is a fix for 5.1.